File: richdoc/richdoc-cli/src/lib/schema.ts
/**
 * Schema for richdoc custom elements.
 *
 * Loaded from `../../../assets/schema.json` (relative to this file), which is
 * produced by `build.ts` from the per-component `*.schema.ts` sources. The
 * CLI never hand-maintains the vocabulary — single source of truth.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

export interface TagSpec {
	required?: readonly string[];
	optional?: readonly string[];
	customChildren?: readonly string[] | "any";
	allowedParents?: readonly string[];
	enums?: Readonly<Record<string, readonly string[]>>;
}

interface SchemaFile {
	$schema?: string;
	generated?: string;
	tags: Record<string, TagSpec>;
}

const SCHEMA_PATH = resolve(
	dirname(fileURLToPath(import.meta.url)),
	"..",
	"..",
	"..",
	"assets",
	"schema.json",
);

function loadSchema(): SchemaFile {
	try {
		const text = readFileSync(SCHEMA_PATH, "utf8");
		const parsed = JSON.parse(text) as SchemaFile;
		if (!parsed || typeof parsed !== "object" || !parsed.tags) {
			throw new Error("schema.json missing 'tags' object");
		}
		return parsed;
	} catch (err) {
		const msg = err instanceof Error ? err.message : String(err);
		throw new Error(
			`Could not load richdoc schema from ${SCHEMA_PATH}: ${msg}. Run \`richdoc build\` (or \`bun run build\` from the richdoc root) to generate it.`,
		);
	}
}

const SCHEMA_FILE = loadSchema();

export const SCHEMA: Readonly<Record<string, TagSpec>> = SCHEMA_FILE.tags;
export const ALLOWED_TAGS: readonly string[] = Object.keys(SCHEMA);
export const SCHEMA_GENERATED_AT: string | undefined = SCHEMA_FILE.generated;
export const SCHEMA_FILE_PATH = SCHEMA_PATH;

/** True if `tag` starts with `rd-`. */
export function isRdTag(tag: string): boolean {
	return tag.toLowerCase().startsWith("rd-");
}

File: richdoc/richdoc-cli/src/richdoc_cli/schema.py
"""Load the rd-* component schema from <framework>/assets/schema.json.

The schema is produced by `pnpm build` from the per-component sources. The
CLI never hand-maintains the vocabulary — single source of truth.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .paths import SCHEMA_PATH


@dataclass(frozen=True)
class SchemaFile:
    tags: dict[str, dict[str, Any]]
    generated: str | None
    path: str


_CACHED: SchemaFile | None = None


def load_schema() -> SchemaFile:
    """Read and cache schema.json. Raises FileNotFoundError-ish errors with a hint."""
    global _CACHED
    if _CACHED is not None:
        return _CACHED

    try:
        text = SCHEMA_PATH.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise SchemaLoadError(
            f"Could not load richdoc schema from {SCHEMA_PATH}: {exc}. "
            "Run `pnpm build` from richdoc/richdoc-lib/ to generate it."
        ) from exc
    except OSError as exc:
        raise SchemaLoadError(
            f"Could not read richdoc schema at {SCHEMA_PATH}: {exc}."
        ) from exc

    try:
        parsed = json.loads(text)
    except json.JSONDecodeError as exc:
        raise SchemaLoadError(
            f"richdoc schema at {SCHEMA_PATH} is not valid JSON: {exc}."
        ) from exc

    tags = parsed.get("tags") if isinstance(parsed, dict) else None
    if not isinstance(tags, dict):
        raise SchemaLoadError(
            f"richdoc schema at {SCHEMA_PATH} is missing a 'tags' object."
        )

    _CACHED = SchemaFile(
        tags=tags,
        generated=parsed.get("generated"),
        path=str(SCHEMA_PATH),
    )
    return _CACHED


class SchemaLoadError(RuntimeError):
    """Raised when the schema can't be loaded — caught by the command layer."""


def is_rd_tag(tag: str) -> bool:
    """True if `tag` starts with `rd-` (case-insensitive)."""
    return tag.lower().startswith("rd-")
